File: src/#328/solution.js
function ListNode(val, next) {
  this.val = val === undefined ? 0 : val
  this.next = next === undefined ? null : next
}

var oddEvenList = function (head) {
  if (head === null) {
    return null
  }

  let current = head
  let oddNode = new ListNode(0, null)
  let res = oddNode
  let evenNode = new ListNode(0, null)
  let evenHead = evenNode

  while (current.next !== null && current.next.next !== null) {
    oddNode.next = current
    oddNode = oddNode.next
    if (current.next !== null) {
      evenNode.next = current.next
      evenNode = evenNode.next
    }
    current = current.next.next
  }

  oddNode.next = current
  oddNode = oddNode.next
  if (current.next !== null) {
    evenNode.next = current.next
    evenNode = evenNode.next
  }

  evenNode.next = null
  oddNode.next = evenHead.next

  return res.next
}

File: src/#328/solution.go
package solution

type ListNode struct {
	Val  int
	Next *ListNode
}

func oddEvenList(head *ListNode) *ListNode {
	if head == nil {
		return nil
	}

	current := head

	oddNode := &ListNode{0, nil}
	res := oddNode
	evenNode := &ListNode{0, nil}
	evenHead := evenNode

	for current.Next != nil && current.Next.Next != nil {
		oddNode.Next = current
		oddNode = oddNode.Next
		if current.Next != nil {
			evenNode.Next = current.Next
			evenNode = evenNode.Next
		}
		current = current.Next.Next
	}

	oddNode.Next = current
	oddNode = oddNode.Next
	if current.Next != nil {
		evenNode.Next = current.Next
		evenNode = evenNode.Next
	}

	evenNode.Next = nil
	oddNode.Next = evenHead.Next

	return res.Next
}
